### PR TITLE
First solver implementation

### PR DESCRIPTION
**Pull request by pnkraemer**
First implementation of a terminal-state EK1. I tried to keep this PR small so we can argue about the design before I throw all the actual solvers at you in separate PRs :)

### CODE REVIEW

File: tornadox/solvers/ek1.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    Uses adaptive steps and PI control.
+    Uses time-varying, scalar-valued diffusion.
+    Uses Taylor-mode initialisation for num_derivatives >= 5,
+    and forward-mode initialisation otherweise.
+
+    Uses a dimension-derivative ordering internally,
+    and does not use projection matrices.
+    """
+
+    EK1State = namedtuple("EK1State", ("u", "dt_proposed", "error_norm", "stats"))
```
Comment (nathanaelbosch) [inline tornadox/solvers/ek1.py:27]: What do you think about making this a 
```python
from typing import NamedTuple

class EK1State(NamedTuple):
    u
    dt_proposed
    error_norm
    stats
```
maybe even outisde of this function? To me that syntax reads much nicer :D

File: tornadox/solvers/ek1.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        u0,
+        rtol,
+        atol,
+        safety=0.95,
+        factor_min=0.2,
+        factor_max=10.0,
+        power_integral_unscaled=0.3,
+        power_proportional_unscaled=0.4,
+    ):
+        del df  # unused
```
Comment (nathanaelbosch) [inline tornadox/solvers/ek1.py:55]: why is this necessary, or helpful?

File: tornadox/solvers/ek1.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        def cond_fun(s):
+            _, state = s
+            return state.error_norm > 1
+
+        @jax.jit
+        def body_fun(s):
+            t, s_prev = s
+
+            # Never exceed the terminal value
+            dt_clipped = jnp.minimum(s_prev.dt_proposed, t1 - t0)
```
Comment (nathanaelbosch) [inline tornadox/solvers/ek1.py:106]: Shouldn't this be
```suggestion
            dt_clipped = jnp.minimum(s_prev.dt_proposed, t1 - t)
```
?

Comment (nathanaelbosch) [reply tornadox/solvers/ek1.py:106]: Ok actually maybe not, `t0` seems to not be what I think it is?

File: tornadox/solvers/ek1.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            s_new = EK1State(
+                u=u_proposed, dt_proposed=dt_new, error_norm=error_norm, stats=stats
+            )
+            return t_new, s_new
+
+        larger_than_1 = 1.1
+        init_val = EK1State(
+            u=state0.u,
+            dt_proposed=state0.dt_proposed,
+            error_norm=larger_than_1,
```
Comment (nathanaelbosch) [inline tornadox/solvers/ek1.py:166]: What is this for?

**Review by nathanaelbosch — COMMENTED** [review]
I had a few comments and questions, but overall things look good and we should be able to resolve this quickly :)

File: tornadox/solvers/ek1.py (hunk 1 above)
Comment (pnkraemer) [reply tornadox/solvers/ek1.py:27]: This is possible. I tend to not use it if I dont want to add types, because it looks a bit naked. But this is personal taste, and if you prefer it then fine :)
(The in-line is also a single line, not 5+2 for spacing, but that is really minor :shrug:)

I would like to keep it inside the function because I don't want this data structure to be "too public", if that makes sense. I already feel bad for using a class here :D

File: tornadox/solvers/ek1.py (hunk 2 above)
Comment (pnkraemer) [reply tornadox/solvers/ek1.py:55]: :shrug:  (I removed it)

File: tornadox/solvers/ek1.py (hunk 3 above)
Comment (pnkraemer) [reply tornadox/solvers/ek1.py:106]: t0 is the point of the previously accepted step.  `t` is actually unused (I made this fact more transparent, thanks for catching this).

File: tornadox/solvers/ek1.py (hunk 4 above)
Comment (pnkraemer) [reply tornadox/solvers/ek1.py:166]: the init_val? jax.lax.while_loop needs consistent types, shapes, etc. for the loop state.
the larger_than_1? Because if this is less than 1, the loop never actually "enters the loop", if that makes sense.

File: tornadox/solvers/ek1.py (hunk 1 above)
Comment (nathanaelbosch) [reply tornadox/solvers/ek1.py:27]: Hmm I see your points. Feel free to decide either way :)

File: tornadox/solvers/ek1.py (hunk 4 above)
Comment (nathanaelbosch) [reply tornadox/solvers/ek1.py:165]: I think I'm just a bit confused with all the body func while loop stuff, but I got it now. Thanks!

**Review by nathanaelbosch — APPROVED** [review]

File: tornadox/solvers/ek1.py (hunk 4 above)
Comment (pnkraemer) [reply tornadox/solvers/ek1.py:165]: Yes this is a little awkward, but cannot be avoided really. Let's keep this in mind and iterate whenever possible :)

File: tornadox/solvers/ek1.py (hunk 1 above)
Comment (pnkraemer) [reply tornadox/solvers/ek1.py:27]: Let's keep it for now and iterate if necessary :)
